internal/provider/string.go

```diff
@@ -257,3 +322,67 @@ func generateRandomBytes(charSet *string, length int) ([]byte, error) {
 func readNil(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return nil
 }
+
+func resourceStateUpgradeAddNumeric(_ context.Context, rawState map[string]interface{}, _ interface{}, resourceName string) func(_ context.Context, rawState map[string]interface{}, _ interface{}) (map[string]interface{}, error) {
+	return func(_ context.Context, rawState map[string]interface{}, _ interface{}) (map[string]interface{}, error) {
+		if rawState == nil {
+			return nil, fmt.Errorf("resource %s state upgrade failed, state is nil", resourceName)
+		}
+
+		number, ok := rawState["number"].(bool)
+		if !ok {
+			return nil, fmt.Errorf("resource %s state upgrade failed, number could not be asserted as bool: %T", resourceName, rawState["number"])
+		}
+
+		rawState["numeric"] = number
+
+		return rawState, nil
+	}
+}
+
+func customDiffIfValue(key string) func(context.Context, *schema.ResourceDiff, interface{}) error {
```

Comment:
I think this function needs a bit of godoc to explain what's needed for. I remember a bit of what you explained verbally, but I will forget in a few hours.

Reply:
Agreed. Added.

Reply:
The naming and single parameter confused me a little bit here at first glance! I wonder if it is worth renaming and making this slightly more generic (not sure how useful it'd be outside this provider, but I think the naming and signature will make it more clear what the intent is), e.g.

```go
func planDefaultIfAllNull(default interface{}, keys ...string) []schema.CustomizeDiffFunc {
  var result []schema.CustomizeDiffFunc

  for _, key := range keys {
    result = append(result, customdiff.IfValue(/*...*/))
  }

  return result
}
```

With the callers then being `planDefaultIfAllNull(true, "number", "numeric")...`